dnn/neural_network.py

```diff
@@ -65,14 +67,20 @@ def __set_costfunc(self, costfunc):
         sys.stdout.write("Error: The lossfunc is not found\n")
         sys.exit(1)
 
-    def forwardpropagation(self, train_data):
+    def forwardpropagation(self, train_data, istrain=False):
         self.z[0][0] = 1
         self.z[0][1:] = train_data
         for i in range(len(self.structure) - 2):
-            self.y[i + 1][1:] = self.weight[i] @ self.z[i]
-            self.z[i + 1] = self.actfunc(self.y[i + 1])
-        self.y[-1] = self.weight[-1] @ self.z[-2]
-        self.z[-1] = self.actfunc(self.y[-1])
+            self.y[i + 1][1:] = self.weight[i] @ self.do[i] @ self.z[i]
+            if istrain:
+                self.z[i + 1] = self.actfunc(self.y[i + 1])
+            else:
+                self.z[i + 1] = self.actfunc(self.y[i + 1]) * (1 - self.dropout[i])
+        self.y[-1] = self.weight[-1] @ self.do[-1] @ self.z[-2]
+        if istrain:
+            self.z[-1] = self.actfunc(self.y[-1])
+        else:
+            self.z[-1] = self.actfunc(self.y[-1]) * (1 - self.dropout[i])
```

Comment:
83  : self.z[-1] = self.actfunc(self.y[-1]) * (1 - self.dropout[i])

self.dropout[i] ?
here is out of for.

Reply:
Thanks!